### DISCUSSION TITLE
Places Where Copilot Can Improve

### DISCUSSION THREAD

**Discussion by davidnaumann-bastian** [community/community#19645] [Copilot Conversations]
Going to start this discussion to talk about what Copilot is currently doing well and what it needs to work on. Updating this/posting more comments to this as I continue to let it into my workflow.

Initial notes:

1. Copilot cannot figure out implementing abstract classes I tried multiple ways such as just let my class be an implementation of the client class and also asking really nicely if Copilot wouldn't mind.
2. This tool works really well as an automated documentation generator. I would almost prefer at this point to disable code suggestions and instead let Copilot run my documentation.
3. It really prefers just generating variables at an increasing complexity (Ex. status, status_timestamp, status_timestamp_format, status_timestamp_format_long, etc.)
4. Copilot (as it was designed) is really good at pulling knowledge to my codebase (really fast to get examples for conversions and surprisingly complex examples on how to design a system)
5. Formatting is pretty good but some of the times Copilot returns with really, really, really bad code

**Comment by D7EAD** [comment]
Github Copilot is great, but we certainly know it is not perfect! During my time with Copilot, from the technical preview stage to production, I've noticed several things with it--many of which are similar to your observations.
- I can certainly agree that Copilot makes documentation very easy. Instead of using something like GhostDoc or another documentation manager to generate documentation, Copilot does this just as well if not better (with some tweaks, of course)!
- Copilot **actively** learns from your existing codebase to provide better suggestions.
   - For instance, if you wanted Copilot to generate something other than a basic class, such as an abstracted or polymorphic class, it would be able to do so much better if it had an existing (or similar) class in its current codebase to be inspired by.
- In terms of Copilot's formatting capacity, in my experience, as described above, it attempts to mimic existing format style, if available. If none is available, or you are attempting to use Copilot to create entirely new, uninspired code, yes, the formatting of it is, frankly, up in the air.
- Variable generation of increasing complexity is an issue I encountered in the preview, but, oddly, I have yet to experience it in production. Then again, I used Copilot for VSCode in preview and VS for production, and, from what I've heard, they don't appear to function the same for everyone.

Overall, I'm very pleased with Copilot so far, despite it sometimes showing its rougher edges!

**Comment by jamesonknutson** [comment]
> This tool works really well as an automated documentation generator. **I would almost prefer at this point to disable code suggestions and instead let Copilot run my documentation.**

[If you use VS Code, I've written an Extension that could (more or less) do exactly this](https://github.com/jamesonknutson/disable-copilot-comment-completions). You could configure the extension to disable inline suggestions in any non-comment textmate scope, and you'd have your wish granted-- it wouldn't throw up suggestions while you're writing your code, but it would provide suggestions whenever you're inside a comment scope (e.g. line comment, block comment, whatever). Perhaps not applicable for you, but maybe it can be of some help.

Tagging the both of you since you both echoed one another's feedback: @D7EAD @davidnaumann-bastian
